xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+    AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub =
+        AggregatedDiscoveryServiceGrpc.newStub(channel);
+    adsStream = new AdsStream(stub);
+    adsStream.start();
+    stopwatch.reset().start();
+  }
+
+  /**
+   * Terminates the RPC connection.
+   */
+  @VisibleForTesting
+  void shutdownRpcStream() {
+    if (adsStream != null) {
+      adsStream.close(Status.CANCELLED.withDescription("shutdown").asException());
+    }
+    if (rpcRetryTimer != null) {
+      rpcRetryTimer.cancel();
+    }
+  }
+
+  /**
+   * Handles LDS response to find the HttpConnectionManager message for the requested resource name.
+   * Proceed with the resolved RouteConfiguration in HttpConnectionManager message of the requested
+   * listener, if exists, to find the VirtualHost configuration for the "xds:" URI
+   * (with the port, if any, stripped off). Or sends an RDS request if configured for dynamic
+   * resolution. The response is NACKed if contains invalid data for gRPC's usage. Otherwise, an
+   * ACK request is sent to management server.
+   */
+  private void handleLdsResponse(DiscoveryResponse ldsResponse) {
+    logger.log(Level.FINE, "Received a LDS response: {0}", ldsResponse);
+    checkState(ldsResourceName != null && configWatcher != null,
+        "No LDS request was ever sent. Management server is doing something wrong");
+    adsStream.ldsRespNonce = ldsResponse.getNonce();
+
+    // Unpack Listener messages.
+    Map<String, Listener> listeners = new HashMap<>(ldsResponse.getResourcesCount());
+    try {
+      for (com.google.protobuf.Any res : ldsResponse.getResourcesList()) {
+        Listener l = res.unpack(Listener.class);
+        listeners.put(l.getName(), l);
+      }
+    } catch (InvalidProtocolBufferException e) {
+      adsStream.sendNackRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName),
+          ldsResponse.getNonce(), "Broken LDS response");
+      configWatcher.onError(Status.fromThrowable(e).augmentDescription("Broken LDS response"));
+      return;
+    }
+
+    // Unpack HttpConnectionManager messages
+    HttpConnectionManager requestedHttpConnManager = null;
+    List<HttpConnectionManager> httpConnectionManagers = new ArrayList<>();
+    try {
+      for (Map.Entry<String, Listener> entry : listeners.entrySet()) {
+        HttpConnectionManager hm =
+            entry.getValue().getApiListener().getApiListener().unpack(HttpConnectionManager.class);
+        httpConnectionManagers.add(hm);
+        if (entry.getKey().equals(ldsResourceName)) {
+          requestedHttpConnManager = hm;
+        }
+      }
+    } catch (InvalidProtocolBufferException e) {
+      adsStream.sendNackRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName),
+          ldsResponse.getNonce(), "Broken LDS response");
+      configWatcher.onError(Status.fromThrowable(e).augmentDescription("Broken LDS response"));
+      return;
+    }
+
+    boolean invalidData = false;
```

Comment:
Instead of boolean consider using a String *AND* a code that you can set and then use in your errorDetail later.

Reply:
Added `Code` in NACK. Currently we put a somewhat general message in error_detail (i.e., "HttpConnectionManager contains invalid information for gRPC's usage"). This gives TD debugging information to narrow down the scope to data validness. This is a tradeoff similar to whether we should have verbose logs that can locate bug precisely or concise logs that gives you some information and requires some investigation. There are many cases for response data being invalid, it would be very verbose to provide highly precise error message about exactly which field is invalid. I believe current solution gives a good leverage.

Reply:
Changed to provide detailed error messages.